### PR TITLE
[WIP] TYCK order

### LINKED ISSUE #40: Tyck order
Opened during Walpurgis night

### PR DESCRIPTION
**Pull request by imkiva**
Will update the description as the development goes.

- [x] fix #40
- [x] #46

### CODE REVIEW

File: base/src/main/java/org/aya/util/MutableGraph.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  private boolean hasPath(@NotNull MutableSet<T> book, @NotNull T from, @NotNull T to) {
+    if (from == to) return true;
+    if (book.contains(from)) return false;
+    for (var test : suc(from)) if (hasPath(book, test, to)) return true;
+    book.add(from);
+    return false;
+  }
+
+  public @NotNull ImmutableSeq<ImmutableSeq<T>> findRing() {
+    return topologicalOrder().filter(scc -> scc.size() > 1);
```
Comment (ice1000) [inline base/src/main/java/org/aya/util/MutableGraph.java:27]:
```suggestion
    return topologicalOrder().filter(scc -> scc.sizeGreaterThan(1));
```

File: base/src/main/java/org/aya/util/MutableGraph.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        SCCs.append(scc.toImmutableSeq());
+      }
+    }
+
+    public ImmutableSeq<ImmutableSeq<T>> tarjan() {
+      // view should be lazy or this code will blow up
+      E.keysView().filter(v -> info(v).noIndex()).forEach(this::makeSCC);
+      // tarjan order is the reverse of topological order
+      return SCCs.toImmutableSeq().reversed();
```
Comment (ice1000) [inline base/src/main/java/org/aya/util/MutableGraph.java:105]:
```suggestion
      return SCCs.reversed().toImmutableSeq();
```

**Review by ice1000 — COMMENTED** [review]
These are just commends don't trust me too much

File: base/src/main/java/org/aya/util/MutableGraph.java (hunk 2 above)
Comment (imkiva) [reply base/src/main/java/org/aya/util/MutableGraph.java:105]: There's no `reversed` in `Buffer` 😭

Comment (imkiva) [reply base/src/main/java/org/aya/util/MutableGraph.java]: Okay. I found one

**Comment by ice1000** [comment]
> There's no `reversed` in `Buffer` 😭 

Then I suggest creating a view, reverse it and collect

**Comment by ice1000** [comment]
Are you sure it's fine to mutate SCCs?

**Comment by ice1000** [comment]
Well I guess so
bors r+

**Comment by imkiva** [comment]
It is unfinished!!!
bors cancel

**Comment by ice1000** [comment]
> It is unfinished!!! bors cancel

I'm sorry

**Comment by imkiva** [comment]
> > It is unfinished!!! bors cancel
> 
> I'm sorry

It's fine. 🤣

**Comment by imkiva** [comment]
Considering in this PR we grouped statements into SCCs and each SCC is independent of each other, we may:

- SCC-level incremental and even real-time compilation for LSP as user types. SCCs are relatively small so it costs less than compiling the whole file.
- Library system needs file dependency info which can be easily made now, thanks to the generalized version of the graph.

**Review by ice1000 — APPROVED** [review]
bors merge

**Comment by ice1000** [comment]
> Build succeeded:
> 
> * gradle-check

Thanks!

**Pull request merged**